### ISSUE TITLE
Parallelism/tactics questions

### ISSUE THREAD

**Issue by ddcc** [Z3Prover/z3#1060]
1. I noticed that parallel threading support was added a few months ago in 37ee4c95c3a2d963002fab22b0f698191d0e92d8 . For the parallel SAT solver, my understanding from reading the code is that the implementation just runs separate SAT solvers in parallel with different random seeds, without sharing any state (e.g. conflict clauses, etc). Is this correct? Do you have any estimates for the average speedup?
2. I also noticed that certain control-flow tactics seems to have had parallel execution support for a while (e.g. `par_and_then` and `par_or`). Is there a plan to switch the default tactics (e.g. the ones used for `QF_BV`) to use these parallel tactics, and likewise what is their performance impact?
3. Is there any documentation on the operation of the default tactics used for e.g. QF_BV, beyond http://rise4fun.com/Z3/tutorial/strategies and `The Strategy Challenge in SMT Solving`, e.g. max_bv_sharing, elim_uncnstr, etc? I'm interested in retrofitting support for incremental operation; but it's unclear to me which tactics T can be applied additively to the decomposition of a set of constraints; e.g. `T({C_0, C_1, ..., C_n}) = T({C_0, C_1, ..., C_n-1}) + T({C_n})`.

**Comment by delcypher** [comment]
@ddcc 
> Is there any documentation on the operation of the default tactics used for e.g. QF_BV, beyond http://rise4fun.com/Z3/tutorial/strategies and The Strategy Challenge in SMT Solving, e.g. max_bv_sharing, elim_uncnstr, etc?

I've only just started learning how to use the tactics but you can find the current implementation of the `qfbv` tactic here.

https://github.com/Z3Prover/z3/blob/master/src/tactic/smtlogics/qfbv_tactic.cpp

It's pretty readable so that's basically the documentation I guess.

For your other questions I'll have to defer to those who know better than me.

**Comment by ddcc** [comment]
> I've only just started learning how to use the tactics but you can find the current implementation of the qfbv tactic here.

> https://github.com/Z3Prover/z3/blob/master/src/tactic/smtlogics/qfbv_tactic.cpp

Yeah, I saw the source code; but I'm curious about the individual tactics that make up the QF_BV tactic.

**Comment by wintersteiger** [comment]
What you can see in qfbv_tactic.cpp is the code that constructs a tactic from other tactics, e.g., and_then, using_params, mk_* all construct the individual tactics). In most cases their names are the same as the names you would use to construct tactics in an SMT2 file, with minor syntactic adjustments. 

The parallel SAT solver is a portfolio, so it runs multiple configurations of a sequential solver in parallel and it exchanges clauses between them; I think that's only units at the moment, @NikolajBjorner knows more. One thing about portfolios is that they become non-deterministic (without explicit measures to prevent it) and there is no way to predict their "usual" behavior, so in terms of performance you'd likely have to run it on your problems and see what average speedup you get. Evaluating this on other problems would probably not give you figures that you can rely on. However, it's often the case that you do get fairly good speedups, in ideal cases even close to linear (see SAT competitions etc). 

par_or and par_and have been there for a long time, but there are currently no plans to use them in the default tactics. And further, there are currently no plans for doing anything in parallel by default, because at this point there are too many users that use Z3 without parallelism and they parallelize on the higher level, as they usually have more than one problem to solve. Running parallel tactics inside of Z3 in this kind of setup wouldn't work very well.

**Comment by ddcc** [comment]
> par_or and par_and have been there for a long time, but there are currently no plans to use them in the default tactics. And further, there are currently no plans for doing anything in parallel by default, because at this point there are too many users that use Z3 without parallelism and they parallelize on the higher level, as they usually have more than one problem to solve. Running parallel tactics inside of Z3 in this kind of setup wouldn't work very well.

I understand that enabling parallelism by default may not be desirable, but at the moment it doesn't seem straightforward to enable even if desired. For the parallel SAT solver, there is a compile-time flag `m_par` and a runtime flag `m_num_parallel` that both need to be set. And on the tactics side, the code needs to be modified from e.g. `and_then` to `par_and_then`, and there is also a runtime paramater `omp_in_parallel()` that is set only if the code is already running in a parallel region. My understanding of this is that a parent function needs to already be running in parallel, so I'm not sure how/where this is enabled.

> What you can see in qfbv_tactic.cpp is the code that constructs a tactic from other tactics, e.g., and_then, using_params, mk_* all construct the individual tactics). In most cases their names are the same as the names you would use to construct tactics in an SMT2 file, with minor syntactic adjustments.

Ok, thanks. Do you know which (if any) of the tactics are additive?

**Comment by wintersteiger** [comment]
That's correct, there is no flag that magically makes everything parallel. This is a lot more complicated than turning `and_then` into `par_and_then` (the latter only uses parallelism if there are multiple sub-goals, which is relatively rare in current applications). 

OpenMP sets the number of threads according to the system configuration, so whenever the code reaches something like

```
#pragma omp parallel for
...
```

it will use as many threads as `omp_get_num_procs()` returns. 

Note that the SAT portfolio is independent of the tactic framework and options like `sat.parallel_threads` do not have an effect on tactics like `par-or` and `par-and-then`. 

In the past, there was also a portfolio built around the SMT solver (as opposed to just SAT), but that is not available anymore (it didn't work all that well anyways) and there may still be references to that on old webpages. Simple SMT portfolios without lemma sharing are easily replaced by the `par-or` tactic however. 

I don't know what kind of additivity you're referring to in your last question, but there is no documentation on particular properties of tactics, so this may be hard to determine.

**Comment by ddcc** [comment]
> I don't know what kind of additivity you're referring to in your last question, but there is no documentation on particular properties of tactics, so this may be hard to determine.

Ok. It sounds like I'll need to dig through the code or try some examples. In order to add support for incremental solving on the tactic solver, I'm trying to determine for which of the tactics I can save the output from the previous computation, and combine with the output on the new constraints incrementally.

**Comment by NikolajBjorner** [comment]
sat.parallel_threads is an early version of portfolio support in the sat solver. 
I suggest to ignore it for now because it will be refined in a different branch.

It will be replaced in a later check-in (4+ months away) by a portfolio approach that is similar to 
other SAT solvers. The opt branch in https://GitHub.com/nikolajbjorner/z3.git contains the development. 
It gives a modest speedup, mainly as it enables the first-finisher to report the verdict.
The threads exchange lemmas, and you will have an option to include local search (to find satisfying assignments) in case this helps. The benefit on smaller formulas is questionable as there is some overhead with spawning the threads, so the sequential version may do very well in many cases. 
par-or or application level portfolio may also be good options depending on your application.

**Comment by ddcc** [comment]
Thanks for the info!
